contracts/InsuranceArbitrator.sol

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@uma/core/contracts/oracle/interfaces/FinderInterface.sol";
+
+/**
+ * @title Insurance Arbitrator Contract
+ * @notice This example implementation allows insurer to issue insurance policy by depositing insured amount,
+ * designating the insured beneficiary and describing insured event. At any time anyone can submit claim that the insured
+ * event has occurred by posting oracle bonding. Insurance Arbitrator resolves the claim through Optimistic Oracle by
+ * passing templated question with insured event description in ancillary data. If the claim is confirmed this contract
+ * automatically pays out insurance coverage to the insured beneficiary. If the claim is rejected policy continues to be
+ * active ready for the subsequent claim attempts.
+ */
+contract InsuranceArbitrator {
+    /******************************************
+     *  STATE VARIABLES AND DATA STRUCTURES   *
+     ******************************************/
+
+    // Stores state and parameters of insurance policy.
+    struct InsurancePolicy {
+        // Claim state preventing simultaneous claim attempts.
+        bool claimInitiated;
+        // Short description of insured event.
+        string insuredEvent;
+        // Beneficiary address eligible for insurance compensation.
+        address insuredAddress;
+        // Denomination token for insurance coverage.
+        address currency;
+        // Amount of insurance coverage.
+        uint256 insuredAmount;
+    }
+
+    // References all active insurance policies by `policyId`.
+    mapping(bytes32 => InsurancePolicy) insurancePolicies;
+
+    // Maps hash of initiated claims to their `policyId`.
+    // This is used in callback function to potentially pay out the beneficiary.
+    mapping(bytes32 => bytes32) insuranceClaims;
+
+    // Oracle proposal bond set to 0.1% of claimed insurance coverage.
+    uint256 constant oracleBondPercentage = 10e15;
+
+    // Optimistic oracle liveness set to 24h.
+    uint256 constant optimisticOracleLivenessTime = 3600 * 24;
+
+    // Price identifier to use when requesting claims through Optimistic Oracle.
+    bytes32 constant priceIdentifier = "YES_OR_NO_QUERY";
+
+    // Template for constructing ancillary data. The claim would insert `insuredEvent` in between when requesting
+    // through Optimistic Oracle.
+    string constant ancillaryDataHead = 'q:"Had the following insured event occurred as of request timestamp: ';
```

Comment:
I feel like the assertion model conceptually works better when referring to the proposal timestamp. I understand that they should be equivalent in this case, and to actually use the proposal timestamp you should technically submit an eventBased modification. So I guess this is still the best way to do it, but just wanted to bring this weirdness up in case you had thought.